src/Symfony/Contracts/HttpClient/HttpClientInterface.php

```diff
@@ -0,0 +1,88 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Contracts\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\TransportExceptionInterface;
+use Symfony\Contracts\HttpClient\Test\HttpClientTest;
+
+/**
+ * Provides flexible methods for requesting HTTP resources.
+ *
+ * @see HttpClientTest for a reference test suite
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ *
+ * @experimental in 1.1
+ */
+interface HttpClientInterface
+{
+    public const OPTIONS_DEFAULTS = [
+        'auth' => null,         // string - a username:password enabling HTTP Basic authentication
+        'query' => [],          // array - associative array of query string values to merge with the request's URL
+        'headers' => [],        // iterable|string[]|string[][] - headers names provided as keys or as part of values
+        'body' => '',           // array|string|resource|\Traversable|\Closure - the callback SHOULD yield a string
+                                //   smaller than the amount requested as argument; the empty string signals EOF; when
+                                //   an array is passed, it is meant as a form payload of field names and values
+        'json' => null,         // array|\JsonSerializable - when set, implementations MUST set the "body" option to
+                                //   the JSON-encoded value and set the "content-type" headers to a JSON-compatible
+                                //   value it is they are not defined - typically "application/json"
+        'user_data' => null,    // mixed - any extra data to attach to the request (scalar, callable, object...) that
+                                //         MUST be available via $response->getInfo('data') - not used internally
+        'max_redirects' => 20,  // int - the maximum number of redirects to follow; a value lower or equal to 0 means
+                                //       redirects should not be followed; "Authorization" and "Cookie" headers MUST
+                                //       NOT follow except for the initial host name
+        'http_version' => null, // string - defaults to the best supported version, typically 1.1 or 2.0
+        'base_uri' => null,     // string - the URI to resolve relative URLs, following rules in RFC 3986, section 2
+        'buffer' => true,       // bool - whether the content of the response should be buffered or not
+        'on_progress' => null,  // callable - called on download/upload at least once per second with arguments
+                                //            int $dlNow, int $dlSize and array $info; any exceptions thrown from the
+                                //            callback MUST abort the request
```

Comment:
What are `$dlNow`, `$dlSize` and `$info` ?

I would imagine that 

 * dlNow: the current chunk size ?
 * dlSize: total of bytes read so far ? (or is it the total size expected ? in which case this can be null since we cannot know in advance for chunked stream)
 * info: same data as the getInfo request method ?

Reply:
dlNow: how many bytes downloaded so far
dlSize: total size; <= 0 means unknown
info: correct

Reply:
Since a body can be 0 sized wouldn't it be better to allow null instead of <= 0 ?

Reply:
I just had a look, `0` is used when the size is unknown with both curl and fopen. From an implementation perspective, passing `null` to notify this looks like a significant effort. I don't think it's worth it...

Reply:
I why also wondering if this method is called even if there is no body on the response ?

Reply:
Yes, this method is called during all the request/response cycle - that includes just after DNS resolution (before the connection is opened), on *upload* progress (to the extent permitted by the driver - ie curl only), while headers arrive, on download progress, and on response completion.

Reply:
I added a note about that.